### DISCUSSION TITLE
Feature Request: Allow passing property getters/setters as parameters

### DISCUSSION THREAD

**Discussion by YairHalberstadt** [dotnet/csharplang#1901] [Language Ideas]
Currently a property cannot be passed by ref, and with good reason: a property is actually syntax sugar for a setter and/or a getter method, and doesn't necessarily correspond to an underlying field.

However it's also relatively painful to create a lambda from the property getter or setter, and to pass that into a method. 

In https://github.com/dotnet/csharplang/issues/1896 we were struggling with the difficulty of writing an extension method `Set` that looks like this: `Foo?.Property.Set(value)`. The problem was that you cannot pass a property by reference.

How about introducing some syntax sugar that looks something like this:

```csharp
public static void Set<T>(set T property, T value )
{
     property = value;
}

public static void SetIfNotX<T>(get set T property, T value, T x )
{
     if(property != x)
          property = value;
}

public static void T GetAfter10Seconds(get T property)
{
     Thread.Sleep(10000);
     return property;
}
```

These three cases would be converted into IL that looked something like this

```csharp
public static void Set<T>(Action<T> set_property, T value )
{
     set_property(value);
}

public static void SetIfNotX<T>((Func<T> get, Action<T> set) property, T value, T x )
{
     if(property.get() != x)
          property.set(value);
}

public static void T GetAfter10Seconds(Func<T> get_property)
{
     Thread.Sleep(10000);
     return get_property();
}
```

Usage would be like this:

```csharp
Set(Foo.Property, value);

SetIfNotX(Foo.Property, value, someOtherValue);

GetAfter10Seconds(Foo.Property);
```

This would be converted into something like
```
Set(Foo.set_Property,value);

SetIfNotX((Foo.get_Property, Foo.set_Property), value, someOtherValue);

GetAfter10Seconds(Foo.get_Property);
```
Note that get_Property and set_Property are methods that already exist in IL.

I agree the syntax might not be the neatest, but that can be discussed.

### Advantages:

Gives a way to pass properties sort of by ref, as well as giving access to property setters/getters inside a method.

### Disadvantages

Would not play very nicely with older C# code or other languages.

**Comment by svick** [comment]
There is a somewhat similar proposal for events: https://github.com/dotnet/csharplang/issues/139.

**Comment by orthoxerox** [comment]
I'm sure this is a duplicate of a ticket by @alrz, but I can't find it on my phone.

**Comment by alrz** [comment]
Yes, this is similar to https://github.com/dotnet/csharplang/issues/84.

**Comment by YairHalberstadt** [comment]
@alrz, @orthoxerox
That would be a great feature to have. However, whilst related, it is very different.

If #84 was implemented it would:

a) turn this feature into syntax sugar, rather than give you the ability to do anything you couldn't in the past.

b) make this feature slightly easier to implement.

This feature therefore should be considered on it's own, and should stand or fall on it's own merit.

**Comment by jnm2** [comment]
> In #1896 we were struggling with the difficulty of writing an extension method Set that looks like this: 

From my point of view, in #1896 we were struggling to see why `obj?.Prop = value;` isn't a more intuitive and consistent option. 😇

But I'd love this proposal for creating delegates straight to accessors.

**Comment by Adnaco** [comment]
Basic intention allow use property as described below and use inout keyword.
void Parse(inout someProperty, string src), the same as void Parse(ref someField, string src)
intentions:
no changes to someProperty if Parse is failed, keep old value.
shorter then: someProperty = Parse(in someProperty, string src);

**Comment by gmit3** [comment]
I find your suggestion quite useful, but I wouldn't introduce a new syntax for this. As all is known at compile time, "ref property_name" could be internally an alias for a tuple with getter and setter delegates.

Does that make any sense?

**Comment by Adnaco** [comment]
Using ref for properties it is great! It makes no difference in calls between fields and properties!
Now only fields is allowed with ref. Please consider access restrictions too. Some properties can be read-only (no setter). Finally, use an alias is acceptable. (an alias for a tuple with getter and setter delegates).
Thank you.

**Comment by theunrepentantgeek** [comment]
> It makes no difference in calls between fields and properties!

Yes it does. 

This was discussed at length when C# was first released, all those many years ago. 

A **ref** parameter isn't passed in at the start of the method and then copied back out at the end of the method, though many people believe that this is the case.

A **ref** parameter provides direct access to the referenced field, allowing immediate modification.

To illustrate, assume we have this simple container class:

``` csharp
public class Person
{
    public string KnownAs;
}
```

We can then write a simple program to test out the behaviour:

```csharp
static void Main(string[] args)
{
    Person alpha = new Person
    {
        KnownAs = "Alpha"
    };

    Console.WriteLine($"i) alpha.KnownAs = {alpha.KnownAs}");
    Modify(ref alpha.KnownAs);
    Console.WriteLine($"iv) alpha.KnownAs = {alpha.KnownAs}");

    Console.ReadLine();

    void Modify(ref string knownAs)
    {
        knownAs = "Secret";
        Console.WriteLine($"ii) alpha.KnownAs = {alpha.KnownAs}");
        Console.WriteLine($"iii) knownAs = {knownAs}");
    }
}
```

What's the output of this program, do you think?

Some folks will read this and expect that `alpha.KnownAs` won't be modified until the call to `Modify()` returns ...

... but the output is actually this:

```
i) alpha.KnownAs = Alpha
ii) alpha.KnownAs = Secret
iii) knownAs = Secret
iv) alpha.KnownAs = Secret
```

The problem comes down to this ...

... if **ref** for a property is _consistent_ with **ref** for a field, then every assignment to the ref parameter would need to set the property ... which would perform very badly, given that properties allow arbitrary code to execute ... and developers would likely raise the poor performance as a bug.

... if **ref** for a property is _inconsistent_ with **ref** for a field, then the property doesn't get modified until the method returns ... which would confound many developers and would repeatedly get raised as a bug.

Given that there is an _extremely simple_ workaround, one that doesn't have any noxious side effects or confounding behaviour ...

**Comment by HaloFour** [comment]
There is a difference in behavior, but I'd argue that it's unlikely to confound.  VB.NET has supported passing properties `ByRef` since 1.0 and it has not been a source of confusion.  I doubt most C#/.NET programmers are even aware of the immediately observable effects of assigning to a `ref` parameter.

https://sharplab.io/#v2:EYLgHgbALANANiAbsGAXEAnArgOwD4CSAtgA4D2GqAzgAQDKAnlagKZECwAUFwLJkAmWOCxoAFDGQDmGAIYdONRWIwBLRDNbKyJFpQY11cLCICCtAjlaTdXJTVtK6WYDR4yVOABQBKewrtKPJ6Gxt4OATQAwmQ4VGTCAHQA6qqsADIeLMEyRixh/ooAojj89M7hfnZOLkEAQgwASiwAZjQqNGY0Fla6+RGK7QC8NACMAEwAzBV2PGM+FcWl1RUV1a5zff3RsfEsyaksGThZIXkLJWXAXIuuAkIsQA===

**Comment by jnm2** [comment]
My gut feeling is that if my code depends on the difference between the two, it's poor coding. I'd be happy to get rid of the connascence.

**Comment by Thaina** [comment]
I don't against the idea but I would like to propose an opposite approach

We might instead have a keyword function `getof` and `setof` and `getsetof` to extract `get`/`set` from the property instead

```C#
Func<SomeValueType> getSomeProperty = getof(obj.SomeProperty);
Action<SomeValueType> setSomeProperty = setof(obj.SomeProperty);
(Func<SomeValueType>,Action<SomeValueType>) getsetSomeProperty = getsetof(obj.SomeProperty);
FuncRef<SomeValueType> getSomeProperty = getof(obj.SomeRefProperty);
```

With this we could just use the same callback code with property

**Comment by skarllot** [comment]
@Thaina maybe `getterof` and `setterof`.

**Reply by siimnelis** [reply]
Is this maybe implemented in newer C# versions? I would love to have something like this to implement State and Strategy pattern. Where the beheviour is encapsulated to a standalone class, but these classes usually need to change some properties on the main class and if i don't want to make main class setters public or internal that matter, i need to use the Action<T> as a parameter and it feels so messy then.
